File: src/lang/ruby/utuclient/lib/utu/stackish.rb
# 
# Utu -- Saving The Internet With Hate
# 
# Copyright (c) Zed A. Shaw 2005 (zedshaw@zedshaw.com)
# 
# This file is modifiable/redistributable under the terms of the GNU
# General Public License.
# 
# You should have recieved a copy of the GNU General Public License along
# with this program; see the file COPYING. If not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 0211-1307, USA.
# 

require 'strscan'


# Augments String with to_blob and is_blob? for tagging
# strings as having BLOB type for stackish.
class String

  # Does nothing more than set @is_blob=true and
  # return the original string (so it's fast).  In
  # order for binary data or other formats to be 
  # encapsulated in a Stackish BLOB you need to do this.
  def to_blob
    self.instance_variable_set(:@is_blob, true)
    self
  end

  # Tells you if a string is a BLOB.
  def is_blob?
    self.instance_variable_get(:@is_blob)
  end

  # Simply runs the parser on this string to convert it into a stackish structure.
  def from_stackish
    Utu::Stackish::Parser.new.process(self)
  end
end

class Hash
  # Processes this hash as a stackish structure to produce a string
  # in stackish format.
  def to_stackish
    Utu::Stackish::Emitter.new.process([self])
  end

  # Makes it so you can use a s/:name to access the
  # contents of the stackish.  Used more with the Array./
  # notation.
  def /(member)
    self[member]
  end
end


class Array
  # Processes this array containing stackish structures into a series
  # of stackish structures as one string.
  def to_stackish
    Utu::Stackish::Emitter.new.process(self)
  end

  # Used to access the members of a stackish struct by the
  # names of the child nodes.  You do it with the divide
  # doing struct/:msg/:child and it can take an Array.
  # When given an array (or anything responding to :collect)
  # it will build the list of all children that match it.
  #
  # This means you can do:  test/:msg/[:chat,:from]/:thing and
  # get all of the :thing children in chat or from.
  def /(member)
     results = if member.respond_to? :collect
                 self.collect {|child|
                   x = member.collect {|sel| child/sel }.flatten.compact
                   x.empty? ? nil : x
                 }
               elsif member.is_a? Fixnum
                 self[member]
               else
                 results = self.collect {|list| 
                   list/member if (list.respond_to? :to_stackish)
                 }.flatten
               end

     results.compact! if results.respond_to? :compact
     results
  end
end


module Utu
  module Stackish
    # Exception thrown by the Parser when it has a problem.
    class ParsingError < StandardError; end

    # Exception thrown by the Emitter when it has a problem.
    class EmittingError < StandardError; end

    # The Parser takes a Stackish string and parses it into a 
    # series of Hashes, Arrays, Strings, Integers, and Floats
    # for you to handle in Ruby.
    class Parser

      # The current partially parsed structure, don't mess with it.
      attr_reader :parse_stack

      # The current marked parts of the stackish structure.
      attr_reader :mark_stack

      # The StringScanner tokens used during parsing.
      TOKENS = [
        [/\s+/ , :space],
        [/\[/ , :mark],
        [/\]/ , :group],
        [/:/, :colon],
        [/'/, :squote],
        [/@/  , :attrib],
        [/(\+|\-)?\d+\.\d+/ , :float],
        [/\d+/ , :number],
        [/\w+(\w|\-|_|\.|:)*/ , :word],
        [/".*?"/  , :string],
      ]

      # Inverse mapping of :token => regex for matching a specific
      # token in the stream.
      TOKENS_INVERTED = TOKENS.inject({}) {|i,v| i[v.last] = v.first; i }  

      # Given a string in Stackish format the process method will
      # parse it and convert it into a mixture of Arrays, Hashes,
      # Strings, Integers, and Floats.  Hashes are used as individual
      # named elements with the key being a Symbol for the word.  Each
      # Hash has an Array of children (which can be empty).  Each
      # child can then be an Array (for unnamed groups), Integer, Float,
      # String, or more Hash.
      #
      # For example, if you give this:
      #
      #   [ test 
      #
      # Then you'll get:
      #
      #   [{:test=>[]}] 
      #
      # Then if you add elements to the test node like so:
      #
      #   [ 1 2 3 4 test
      #
      # Then you'll get:
      #
      #   [{:test=>[1, 2, 3, 4]}]
      #
      # (Notice the numbers are "backwards").  If you add a group into that:
      #
      #   [ [ "hi" ] 1 2 3 4 test
      #
      # Then you'll get:
      #
      #   [{:test=>[["hi"], 1, 2, 3, 4]}]
      #
      # And if you decide to name this "hi" group:
      #
      #   [ [ "hi" howdy 1 2 3 4 test
      #
      # You'll then get:
      #
      #   [{:test=>[{:howdy=>["hi"]}, 1, 2, 3, 4]}]
      # 
      # And finally, you use the String#to_blob and String#is_blob? functions
      # to create blobs.
      #
      # The most convenient way to use Stackish is to simply work with String#from_stackish
      # and Array#to_stackish (or Hash#to_stackish).
      #
      # NOTE:  Currently process will throw in exception if the string doesn't 
      # represent a complete stackish structure. 
      def process(string)
        @parse_stack = []
        @mark_stack = []
        @scanner = StringScanner.new(string)

        tok = ""
        while !@scanner.eos?
          TOKENS.each do |p,r|
            tok = @scanner.check(p)
            if tok and tok.length > 0
              skip(tok.length)
              # skip spaces but process everything else
              self.send(r, tok) unless r == :space
              break
            end
          end

          if !tok or tok.length == 0
            raise ParsingError.new("Invalid stackish string at #{@scanner.pos} character, failed to parse.")
          end
        end

        raise ParsingError.new("Incomplete Stackish structure in string.  Expecting at least #{@mark_stack.length} words or groups.") if !finished?

        return @parse_stack
      end

      # Since Stackish is stack ordered, you can tell when you've processed
      # all of the structure very easily.
      def finished?
        @mark_stack.empty?
      end

      protected

      # Takes the current @parse_stack stack top position and pushes it onto
      # the @mark_stack stack for later popping off to produce groups.
      def mark(token)
        @mark_stack.push(@parse_stack.length)
      end

      # Creates an unnamed group (doesn't use token)
      def group(token)
        @parse_stack << collapse
      end

      # Produces a string (NOT blob) by stripping off the \" chars around
      # what scanner produces.
      def string(token)
        # TODO: gotta be a way to make strscan do this from the regex
        @parse_stack << token[1 ... token.length-1]
      end

      # Process a BLOB specification of the format 'LENGTH:DATA'.
      def squote(token)
        size = match(:number).to_i
        match(:colon)
        blob = @scanner.rest[0 ... size].to_blob
        skip(size)
        match(:squote)
        @parse_stack << blob
      end

      # Names whatever is on the top with the given attribute word
      def attrib(token)
        @parse_stack << name((token + match(:word)), @parse_stack.pop)
      end

      def float(token)
        @parse_stack << token.to_f
      end

      def number(token)
        @parse_stack << token.to_i
      end

      # Takes all the elements from the current top to the most recently
      # pushed mark, collapses them into a group, then names that group.
      def word(token)
        @parse_stack << name(token, collapse)
      end

      # Matches the given token and returns it.  Raises a ParsingError if the
      # match doesn't happen.
      def match(name)
        m = @scanner.scan(TOKENS_INVERTED[name])
        if !m or m.length == 0
          raise ParsingError.new("Expected #{name} token at #{@scanner.pos} character of input string.")
        end
        return m
      end

      # Skips over this much of the input text.
      def skip(length)
        begin
          @scanner.pos += length
        rescue RangeError
          raise ParsingError.new("Not enough input in string at #{@scanner.pos}, expecting #{@scanner.pos + length - @scanner.string.length} more.")
        end
      end

      # Names a group with a Symbol by wrapping it with a single
      # name=>children mapping Hash.
      def name(name, element)
        { name.to_sym => element}
      end

      # Takes the most recent mark off the @mark_stack stack, slices from that point
      # on into a separate array, then returns it so this new group can be
      # processed.
      def collapse
        top = @mark_stack.pop
        raise ParsingError.new("Unbalanced stackish structure at #{@scanner.pos}.") unless top
        @parse_stack.slice!(top, @parse_stack.length)
      end
    end


    class Emitter
      def process(stackish)
        stackish.inject([]) {|a,el| a << convert(el); a }.join(" ") + " \n"
      end

      def convert(element, named=false)
        case element
        when Hash
          name = element.keys[0].to_s
          [convert(element.values[0], name[0] != "@"[0])] + [name]
        when Array
          ["["] + element.inject([]) {|i,x| i << convert(x); i } + (named ? [] : [']'])
        when String
          if element.is_blob?
            "'#{element.length}:#{element.to_s}'"
          else
            "\"#{element.to_s}\""
          end
        when Integer
          element.to_s
        when Float
          element.to_s
        else
          raise EmittingError.new("Unknown Stackish type #{element.class}: #{element.inspect}")
        end
      end
    end
  end
end

File: src/stackish/stackish.c
#line 1 "stackish/stackish.rl"
/*
 * Utu -- Saving The Internet With Hate
 *
 * Copyright (c) Zed A. Shaw 2005 (zedshaw@zedshaw.com)
 *
 * This file is modifiable/redistributable under the terms of the GNU
 * General Public License.
 *
 * You should have recieved a copy of the GNU General Public License along
 * with this program; see the file COPYING. If not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 0211-1307, USA.
 */

#include "stackish/ragel.h"
#include "stackish/stackish.h"
#include <myriad/defend.h>

#define push(T,M,F) handle_push(parser, TYPE_##T, PTR_TO(M), LEN(M, F)) 

inline int handle_push(stackish_parser *parser, enum NodeType type, const char *start, size_t length)
{
  assert_not(parser, NULL);
  assert_not(start, NULL);

  Node *current = parser->current;

  check(current, "parsing failure, no current node");
  check(parser->root, "parsing failure, no root node");
  check(current->type == TYPE_GROUP, "parsing failure, pushing onto a node that's not a group");

  Node *node = Node_from_str(current, type, start, length);

  return node != NULL;
  on_fail(return 0);
}


/** Returns false if the document isn't done, otherwise true and -1 if there's a failure.*/
inline int handle_word(stackish_parser *parser, const char *start, size_t length)
{
  assert_not(parser, NULL);

  Node *current = parser->current;

  check(parser->root, "parsing error, root node not ready");
  check(parser->current, "parsing error, parser doesn't have an active current node");

  if(start == NULL) {
    if(current->name != NULL) bdestroy(current->name);
    current->name = NULL;
  } else {
    Node_name(current, blk2bstr(start, length));
  }

  // either quit if we're at the root or move up to the parent
  if(current == parser->root) {
    return 1;
  } else {
    parser->current = current->parent;
    return 0;
  }

  on_fail(return -1);
}


/** Similar to handle_word this returns true when the document is done, false
  otherwise. */
inline int handle_group(stackish_parser *parser) 
{
  return handle_word(parser, NULL, 0);
}


inline int handle_attr(stackish_parser *parser, const char *start, size_t length)
{
  assert_not(parser, NULL);

  Node *current = parser->current->child;
  check(current, "parsing failure, attempting to set an attribute of a node with no children");

  Node_name(current, blk2bstr(start, length));

  return 1;
  on_fail(return 0);
}


inline void handle_start(stackish_parser *parser)
{
  assert_not(parser, NULL);

  // new group will ignore root if it is NULL, and attach it otherwise
  Node *mark = Node_new_group(parser->current);

  // first node, so set the root and current to it
  if(parser->root == NULL) {
    parser->root = mark;
  } 

  // now we set the current to the emark
  parser->current = mark;
}


/** machine **/
#line 160 "stackish/stackish.rl"


/** Data **/

#line 114 "stackish/stackish.c"
static const char _stackish_parser_actions[] = {
	0, 1, 0, 1, 1, 1, 2, 1, 
	3, 1, 4, 1, 5, 1, 6, 1, 
	7, 1, 8, 1, 9, 2, 0, 3, 
	2, 1, 0, 2, 1, 4, 2, 1, 
	7, 2, 2, 0, 2, 2, 4, 2, 
	2, 7, 2, 6, 0, 2, 6, 4, 
	2, 6, 7, 2, 9, 0, 2, 9, 
	4, 2, 9, 7
};

static const char _stackish_parser_key_offsets[] = {
	0, 0, 1, 2, 4, 7, 8, 10, 
	13, 15, 19, 35, 51, 69, 87
};

static const char _stackish_parser_trans_keys[] = {
	34, 34, 48, 57, 58, 48, 57, 39, 
	48, 57, 46, 48, 57, 48, 57, 65, 
	90, 97, 122, 32, 34, 39, 43, 45, 
	64, 91, 93, 9, 13, 48, 57, 65, 
	90, 97, 122, 32, 34, 39, 43, 45, 
	64, 91, 93, 9, 13, 48, 57, 65, 
	90, 97, 122, 32, 34, 39, 43, 64, 
	91, 93, 95, 9, 13, 45, 46, 48, 
	58, 65, 90, 97, 122, 32, 34, 39, 
	43, 64, 91, 93, 95, 9, 13, 45, 
	46, 48, 58, 65, 90, 97, 122, 32, 
	34, 39, 43, 45, 46, 64, 91, 93, 
	9, 13, 48, 57, 65, 90, 97, 122, 
	0
};

static const char _stackish_parser_single_lengths[] = {
	0, 1, 1, 0, 1, 1, 0, 1, 
	0, 0, 8, 8, 8, 8, 9
};

static const char _stackish_parser_range_lengths[] = {
	0, 0, 0, 1, 1, 0, 1, 1, 
	1, 2, 4, 4, 5, 5, 4
};

static const char _stackish_parser_index_offsets[] = {
	0, 0, 2, 4, 6, 9, 11, 13, 
	16, 18, 21, 34, 47, 61, 75
};

static const char _stackish_parser_trans_targs_wi[] = {
	10, 2, 10, 2, 4, 0, 5, 4, 
	0, 10, 0, 7, 0, 8, 7, 0, 
	11, 0, 12, 12, 0, 10, 1, 3, 
	6, 6, 9, 10, 10, 10, 14, 13, 
	13, 0, 10, 1, 3, 6, 6, 9, 
	10, 10, 10, 11, 13, 13, 0, 10, 
	1, 3, 6, 9, 10, 10, 12, 10, 
	12, 12, 12, 12, 0, 10, 1, 3, 
	6, 9, 10, 10, 13, 10, 13, 13, 
	13, 13, 0, 10, 1, 3, 6, 6, 
	8, 9, 10, 10, 10, 14, 13, 13, 
	0, 0
};

static const char _stackish_parser_trans_actions_wi[] = {
	21, 1, 7, 0, 1, 0, 17, 0, 
	0, 11, 0, 0, 0, 0, 0, 0, 
	0, 0, 0, 0, 0, 0, 0, 0, 
	1, 1, 1, 9, 15, 0, 1, 1, 
	1, 0, 5, 5, 5, 33, 33, 33, 
	36, 39, 5, 0, 33, 33, 0, 19, 
	19, 19, 51, 51, 54, 57, 0, 19, 
	0, 0, 0, 0, 0, 13, 13, 13, 
	42, 42, 45, 48, 0, 13, 0, 0, 
	0, 0, 0, 3, 3, 3, 24, 24, 
	0, 24, 27, 30, 3, 0, 24, 24, 
	0, 0
};

static const int stackish_parser_start = 10;
static const int stackish_parser_first_final = 10;
static const int stackish_parser_error = 0;

static const int stackish_parser_en_main = 10;

#line 164 "stackish/stackish.rl"

RAGEL_INIT(stackish_parser, {
    
#line 203 "stackish/stackish.c"
	{
	cs = stackish_parser_start;
	}
#line 167 "stackish/stackish.rl"
})

RAGEL_DEFINE_FUNCTIONS(stackish_parser, {
    
#line 212 "stackish/stackish.c"
	{
	int _klen;
	unsigned int _trans;
	const char *_acts;
	unsigned int _nacts;
	const char *_keys;

	if ( p == pe )
		goto _out;
	if ( cs == 0 )
		goto _out;
_resume:
	_keys = _stackish_parser_trans_keys + _stackish_parser_key_offsets[cs];
	_trans = _stackish_parser_index_offsets[cs];

	_klen = _stackish_parser_single_lengths[cs];
	if ( _klen > 0 ) {
		const char *_lower = _keys;
		const char *_mid;
		const char *_upper = _keys + _klen - 1;
		while (1) {
			if ( _upper < _lower )
				break;

			_mid = _lower + ((_upper-_lower) >> 1);
			if ( (*p) < *_mid )
				_upper = _mid - 1;
			else if ( (*p) > *_mid )
				_lower = _mid + 1;
			else {
				_trans += (_mid - _keys);
				goto _match;
			}
		}
		_keys += _klen;
		_trans += _klen;
	}

	_klen = _stackish_parser_range_lengths[cs];
	if ( _klen > 0 ) {
		const char *_lower = _keys;
		const char *_mid;
		const char *_upper = _keys + (_klen<<1) - 2;
		while (1) {
			if ( _upper < _lower )
				break;

			_mid = _lower + (((_upper-_lower) >> 1) & ~1);
			if ( (*p) < _mid[0] )
				_upper = _mid - 2;
			else if ( (*p) > _mid[1] )
				_lower = _mid + 2;
			else {
				_trans += ((_mid - _keys)>>1);
				goto _match;
			}
		}
		_trans += _klen;
	}

_match:
	cs = _stackish_parser_trans_targs_wi[_trans];

	if ( _stackish_parser_trans_actions_wi[_trans] == 0 )
		goto _again;

	_acts = _stackish_parser_actions + _stackish_parser_trans_actions_wi[_trans];
	_nacts = (unsigned int) *_acts++;
	while ( _nacts-- > 0 )
	{
		switch ( *_acts++ )
		{
	case 0:
#line 110 "stackish/stackish.rl"
	{ MARK(mark, p); }
	break;
	case 1:
#line 111 "stackish/stackish.rl"
	{ if(!push(NUMBER, mark, p)) goto _out; }
	break;
	case 2:
#line 112 "stackish/stackish.rl"
	{ if(!push(FLOAT, mark, p)) goto _out; }
	break;
	case 3:
#line 113 "stackish/stackish.rl"
	{ if(!push(STRING, mark, p)) goto _out; }
	break;
	case 4:
#line 114 "stackish/stackish.rl"
	{ handle_start(parser); }
	break;
	case 5:
#line 115 "stackish/stackish.rl"
	{ if(!push(BLOB, mark, p)) goto _out; else parser->more = 0;}
	break;
	case 6:
#line 116 "stackish/stackish.rl"
	{
    if(handle_word(parser, PTR_TO(mark), LEN(mark, p))) {
      // all done, stop processing
      goto _out;
    }
  }
	break;
	case 7:
#line 122 "stackish/stackish.rl"
	{
    if(handle_group(parser)) {
      // all done, stop processing
      goto _out;
    }
  }
	break;
	case 8:
#line 128 "stackish/stackish.rl"
	{ 
    char *end = NULL; 
    parser->more = strtoul(PTR_TO(mark), &end, 10); 
    assert(end == p && "buffer overflow processing blob length");
    MARK(mark, p+1);
    if(parser->more > 0) {
      // we only need to break if there needs to be a chunk to process
      goto _out;
    }
  }
	break;
	case 9:
#line 138 "stackish/stackish.rl"
	{ if(!handle_attr(parser, PTR_TO(mark), LEN(mark, p))) goto _out; }
	break;
#line 344 "stackish/stackish.c"
		}
	}

_again:
	if ( cs == 0 )
		goto _out;
	if ( ++p != pe )
		goto _resume;
	_out: {}
	}
#line 171 "stackish/stackish.rl"
    }, 
    {
    
#line 359 "stackish/stackish.c"
#line 174 "stackish/stackish.rl"
    });


void stackish_node_clear(stackish_parser *parser) {
  assert_not(parser, NULL);

  if(parser->root) {
    Node_destroy(parser->root);
    parser->root = NULL;
    parser->current = NULL;
  }
}

void stackish_node_destroy(stackish_parser *parser) {
  if(parser->root != NULL) Node_destroy(parser->root);
}

